drafts/zip-whitelisted-transparent-programs.rst

```diff
@@ -0,0 +1,282 @@
+  is because the signature verification algorithm [#redjubjub]_ multiplies through by the
+  cofactor before the identity check (which ensures that the single and batch verification
+  algorithms agree on the set of valid signatures). An adversary can modify the signature
+  by adding a nonzero point of small order to R, leaving the signature valid but altering
+  its binary representation.
+
+  - TODO: Upon further investigation, it looks like RedJubjub is not actually malleable
+    for the above reason. We are analysing this and will update the proposal when we are
+    certain. It is also possible that Ed25519 might also be non-malleable for a similar
+    reason, but verifying this would require very careful inspection of libsodium; it is
+    simpler to just exclude them.
+
+Shielded programmability and a full transaction ID malleability fix would both require
+significant changes across the ecosystem. In the interim, we propose in this ZIP the
+simplest possible change to the consensus rules that addresses the above concerns with a
+minimal impact on the existing ecosystem.
+
+
+Conventions
+===========
+
+We use the following notation as defined in the Zcash protocol specification
+[#spec-notation]_, reproduced here for convenience:
+
+- length(*S*) means the length of (number of elements in) *S*.
+
+- { *a* .. *b* } means the set or type of integers from *a* through *b* inclusive.
+
+- *a* || *b* means the concatenation of sequences *a* then *b*.
+
+We reproduce the following conversion function from ZIP 32 [#zip-0032]_:
+
+- I2LEOSP\ :sub:`l`\ (*k*) is the byte sequence *S* of length *l*/8 representing in little-endian order the
+  integer *k* in range {0..2\ :sup:`l`\ -1}.
+
+We also define the following helper notation for scripts:
+
+- PUSHDATA(*S*) is a data push of the byte sequence *S* using the standard script data
+  push encoding:
+
+  - I2LEOSP\ :sub:`8`\ (length(*S*)) || *S* if length(*S*) < ``PUSHDATA1``.
+  - ``PUSHDATA1`` || I2LEOSP\ :sub:`8`\ (length(*S*)) || *S* if length(*S*) <= ``0xff``.
+  - ``PUSHDATA2`` || I2LEOSP\ :sub:`16`\ (length(*S*)) || *S* if length(*S*) <= ``0xffff``.
+  - ``PUSHDATA4`` || I2LEOSP\ :sub:`32`\ (length(*S*)) || *S* otherwise.
+
+
+Specification
+=============
+
+Transparent program
+-------------------
+
+A transparent program is defined by four properties:
+
+- A 1-byte program type.
+- A byte sequence ``predicate`` containing a prefix-free encoding describing the
+  predicate.
+- A non-malleable byte sequence ``witness`` containing a prefix-free encoding of evidence
+  satisfying the predicate.
+- A deterministic verification algorithm ``verify_program`` that takes as arguments
+  ``predicate``, ``witness``, and a context object providing deterministic public
+  information about the transaction as well as the block chain (up to and including the
+  block that the transaction is mined in). It returns ``true`` if the predicate is
```

Comment:
Currently, while a tx can become invalid (if an input is spent, or if it expires), a Script never can. This means Script validity can be evaluated and cached as soon as a tx is received. It would be nice to preserve this property for WTPs. Passing in the blockchain as part of the evaluation context prevents us from doing this though

Reply:
Hmm, this is an interesting point. The reason for the context object was to enable logic like relative timelocks to just be part of the WTP. The way in which `OP_CLTV` and `OP_CSV` are implemented to ensure the Script never becomes invalid, is to have them operate on public transaction fields (`nLocktime` and `nSequence`), and then have those fields be contextually checked in the consensus rules.

The equivalent for WTPs would be to have an `auxiliary` byte sequence alongside `witness`, and two verification algorithms:

- `verify_auxiliary(auxiliary, context)` - contextual, but requires minimal computation.
- `verify_program(predicate, auxiliary, witness)` - potentially computation-heavy, but deterministic and cacheable.

The question then is whether it is okay for `auxiliary` to not be covered by the sighash (both `nLocktime` and `nSequence` are covered in `SIGHASH_ALL` mode, but the only place we can put arbitrary data is `scriptSig` which is never covered). I don't think it is, which means we'd need to commit to `auxiliary` somewhere. Given that we have not yet pulled in `OP_CSV`, we _could_ repurpose `nSequence` ourselves, and have it contain a small digest of `auxiliary` (probably computed using BLAKE2b-32 for consistency with the rest of the protocol, though maybe that's overkill). Then the consensus rules can enforce `nSequence` contain this digest for transactions containing `OP_PROGRAM`.

Thoughts?

Reply:
Nothing prevents an implementation of the WTP verification from caching anything that it can validly cache. That is, it's possible to split WTP verification into noncontextual and contextual parts, just as an implemenation detail, without any change to the transaction format.

Reply:
> without any complication of the transaction format.

This is very important to me as well :)

> The reason for the context object was to enable logic like relative timelocks to just be part of the WTP.

I think that we could still do this elegantly by passing a context containing only limited information. For example, we could pass in the most recent block height, and the age of the transaction's parents, without passing in the entire history